File: Algorithms/p147_Insertion_Sort_List/p147_Insertion_Sort_List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* insertionSortList(ListNode* head) {
        if(!head) return 0;
        ListNode** p_head = &head;
        ListNode* p1 = head->next;
        ListNode* p2;
        ListNode** pp;
        head->next = 0;
        while(p1){
            pp = p_head;
            while((*pp)&&(*pp)->val<p1->val) pp = &(*pp)->next;
            p2 = p1->next;
            p1->next = *pp;
            *pp = p1;
            p1 = p2;
        }
        return *p_head;
    }
};

File: Algorithms/p147_Insertion_Sort_List/p147_Insertion_Sort_List.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def insertionSortList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if not head:
            return None
        head0 = ListNode(None)
        head0.next = head
        p1 = head.next
        head.next = None
        while p1:
            p = head0
            while p.next and p.next.val<p1.val:
                p = p.next
            p2 = p1.next
            p1.next = p.next
            p.next = p1
            p1 = p2
        return head0.next
